### PR TITLE
CASSANDRA-19985: Enhance CQLSH to support machine-readable output formatting

### PR DESCRIPTION
**Pull request by arvindKandpal-ksolves**
Enhance CQLSH to support machine-readable output formatting (csv, json)

Currently, `cqlsh` output formatting provides tabular formatting designed for human readability, which complicates machine processing in automated pipelines. This patch introduces a `--mode` argument to switch between output formats.

Changes included in this patch:
* Added `--mode` argument supporting `tabular`, `csv`, and `json` formats.
* Retained `tabular` as the default mode to ensure no breaking changes for existing users.
* Disabled ANSI color codes and extra footer text (like `(X rows)`) when `csv` or `json` mode is selected to ensure purely machine-readable output.
* [cite_start]Updated `cqlsh.adoc` documentation and `cqlshrc.sample` configuration file[cite: 1, 2].
* [cite_start]Added unit tests in `test_cqlsh_output.py` to verify both CSV and JSON formats[cite: 1, 2].

patch by Arvind Kandpal; reviewed by <Reviewers> for CASSANDRA-19985

The [Cassandra Jira](CASSANDRA-19985)

### CODE REVIEW

File: pylib/cqlshlib/cqlshmain.py [hunk 1]
```diff
@@ -1009,7 +1016,11 @@ def print_static_result(self, result, table_meta, with_header, tty, row_count_of
 
         formatted_values = [list(map(self.myformat_value, [row[c] for c in column_names], cql_types)) for row in result.current_rows]
 
-        if self.expand_enabled:
+        if self.mode == 'csv':
```
Comment (bschoening) [inline pylib/cqlshlib/cqlshmain.py:1019]: Output formatting should probably move to displaying.py instead of cqlshmain.  Rather than if-then-else, it would probably make sense to create a class like TablePrinter with subclasses for TabularTablePrinter, JsonTablePrinter, CsvTablePrinter.

File: pylib/cqlshlib/test/test_cqlsh_output.py [hunk 2]
```diff
@@ -1017,3 +1017,35 @@ def test_quoted_output_text_in_udts(self):
                                         tty=False, input=query)
         self.assertEqual(0, result)
         self.assertEqual(output.splitlines()[3].strip(), "{data: 'I''m newb'}")
+
+    def test_csv_output(self):
```
Comment (bschoening) [inline pylib/cqlshlib/test/test_cqlsh_output.py:1021]: Unit tests should include all data types and include corner cases.  

- How is the number 10,000 formatted in CVS? The name "Smith, Joe" in CSV?
- The set "{a, b, c}" in JSON?

**Review by bschoening — CHANGES_REQUESTED** [review]

File: pylib/cqlshlib/displaying.py [hunk 3]
```diff
@@ -126,3 +126,86 @@ def color_ljust(self, width, fill=' '):
...
+        self._colnames = [n.strval for n in formatted_names]
+        self._shell.writeresult('[')
+
+    def print_rows(self, formatted_names, formatted_values):
+        import json
+        if formatted_values is None:
+            return
+        for row in formatted_values:
+            row_dict = {self._colnames[i]: col.strval for i, col in enumerate(row)}
+            serialized = json.dumps(row_dict)
```
Comment (bschoening) [inline pylib/cqlshlib/displaying.py:203]: Check Cassandra types UUID, Decimal, or LocalDate in JSON and ensure they have corresponding unit tests. Strings in formatted values may require escaping.

File: pylib/cqlshlib/displaying.py [hunk 4]
```diff
@@ -126,3 +126,86 @@ def color_ljust(self, width, fill=' '):
...
 
 NO_COLOR_MAP = dict()
+
+class TablePrinter:
+    def print_header(self, formatted_names):
+        raise NotImplementedError
+
+    def print_rows(self, formatted_names, formatted_values):
+        raise NotImplementedError
+
```
Comment (bschoening) [inline pylib/cqlshlib/displaying.py:136]: add a static factory method:

@staticmethod
    def factory(format_type, output=sys.stdout):
        """Centralized factory method to return the correct printer."""
        format_map = {
            'csv':   CSVPrinter,
            'json':  JSONPrinter,
            'table': TabularTablePrinter,  # Default ASCII table
        }
        
        # Default to GridPrinter if format is unknown
        printer_cls = format_map.get(format_type.lower(), TabularTablePrinter)
        return printer_cls(output)

File: pylib/cqlshlib/cqlshmain.py [hunk 5]
```diff
@@ -956,32 +962,37 @@ def perform_simple_statement(self, statement):
     def print_result(self, result, table_meta):
         self.decoding_errors = []
 
-        self.writeresult("")
+        if self.mode == 'csv':
```
Comment (bschoening) [inline pylib/cqlshlib/cqlshmain.py:965]: see comment below in TablePrinter, this should be moved into a factory method

File: pylib/cqlshlib/cqlshmain.py [hunk 6]
```diff
@@ -990,15 +1001,18 @@ def print_all(result, table_meta, tty):
...
         if not result.column_names and not table_meta:
             return
 
         column_names = result.column_names or list(table_meta.columns.keys())
         formatted_names = [self.myformat_colname(name, table_meta) for name in column_names]
+
         if not result.current_rows:
-            # print header only
-            self.print_formatted_result(formatted_names, None, with_header=True, tty=tty)
+            if printer is None or isinstance(printer, TabularTablePrinter):
```
Comment (bschoening) [inline pylib/cqlshlib/cqlshmain.py:1012]: probably shouldn't allow printer to be undefined.  Couldn't this be delegated TabularTablePrinter.print_header()?

**Review by bschoening — CHANGES_REQUESTED** [review]

File: pylib/cqlshlib/displaying.py [hunk 7]
```diff
@@ -126,3 +126,86 @@ def color_ljust(self, width, fill=' '):
...
+            self._writer.writerow(self._colnames)
+            self._header_written = True
+        if formatted_values is None:
+            return
+        for row in formatted_values:
+            self._writer.writerow([col.strval for col in row])
+
+    def finish(self):
+        pass
+
```
Comment (bschoening) [inline pylib/cqlshlib/displaying.py:202]: Gemini suggests this simplified approach, check if it works

class CSVPrinter(TablePrinter):
    def print_results(self, colnames, rows):
        writer = csv.writer(self.output)
        writer.writerow(colnames)
        writer.writerows(rows)

class JSONPrinter(TablePrinter):
    def print_results(self, colnames, rows):
        data = [dict(zip(colnames, row)) for row in rows]
        json.dump(data, self.output, indent=4, default=str)
        self.output.write('\n')

File: pylib/cqlshlib/cqlshmain.py [hunk 8]
```diff
@@ -956,32 +963,33 @@ def perform_simple_statement(self, statement):
...
                     break
                 is_first = False
             return num_rows
 
-        num_rows = print_all(result, table_meta, self.tty)
-        self.writeresult("(%d rows)" % num_rows)
+        num_rows = print_all(result, table_meta, self.tty, printer)
+        printer.finish()
+        if self.mode not in ('csv', 'json'):
+            self.writeresult("(%d rows)" % num_rows)
```
Comment (bschoening) [inline pylib/cqlshlib/cqlshmain.py:1046]: should num_rows be inside the if self.mode?

**Comment by arvindKandpal-ksolves** [comment]
Hi @bschoening , Can you look again into it ?

**Comment by arvindKandpal-ksolves** [comment]
cc @bschoening

File: pylib/cqlshlib/displaying.py (hunk 7 above)
Comment (arvindKandpal-ksolves) [reply pylib/cqlshlib/displaying.py:202]: I tested this Gemini-suggested simplified approach, but it unfortunately breaks when Cassandra's paging is involved. 

Because `print_rows` is called multiple times as Cassandra fetches data in chunks (pages), using `json.dump(data, ...)` in `print_rows` causes it to output multiple separate JSON arrays back-to-back. This results in an invalid JSON stream. 

For example, testing this with `PAGING 2` yields the following invalid output:
```json
Connected to Test Cluster at 127.0.0.1:9042
[cqlsh 6.3.0 | Cassandra 7.0-SNAPSHOT | CQL spec 3.4.8 | Native protocol v5]
Use HELP for help.
cqlsh> PAGING 2
Page size: 2
cqlsh> SELECT keyspace_name FROM system_schema.keyspaces;
[
    {
        "keyspace_name": "system_cluster_metadata"
    },
    {
        "keyspace_name": "test"
    }
]
[
    {
        "keyspace_name": "system_auth"
    },
    {
        "keyspace_name": "system_schema"
    }
]
[
    {
        "keyspace_name": "test_bugs"
    },
    {
        "keyspace_name": "system_distributed"
    }
]
[
    {
        "keyspace_name": "system"
    },
    {
        "keyspace_name": "system_traces"
    }
]
cqlsh> 
```
My current streaming approach—printing the opening `[` in `print_header`, handling commas row-by-row, and closing with `]` in `finish()`—ensures we output a single, perfectly valid JSON array regardless of how many pages are fetched, while keeping memory usage strictly low for large `SELECT *` queries.

File: pylib/cqlshlib/cqlshmain.py (hunk 8 above)
Comment (arvindKandpal-ksolves) [reply pylib/cqlshlib/cqlshmain.py:1046]: I tested moving `num_rows = print_all(...)` inside the `if` block, but it actually breaks the machine-readable output. 

Because `print_all()` is the function that actually fetches and prints the data, putting it inside the `if self.mode not in ('csv', 'json')` block means `cqlsh` skips fetching data entirely for CSV/JSON modes (it just outputs an empty `]`). 

We have to execute `print_all()` regardless of the mode to fetch the results, and we just use the `if` block afterward to suppress the final `(X rows)` summary text so it doesn't corrupt the output file.

**Comment by arvindKandpal-ksolves** [comment]
Hi @bschoening , I have tested these suggestions but didn't get the correct result.

**Comment by bschoening** [comment]
@arvindKandpal-ksolves thanks for the effort, I'll try to review this week.

File: pylib/cqlshlib/displaying.py [hunk 9]
```diff
@@ -126,3 +126,101 @@ def color_ljust(self, width, fill=' '):
...
+class JsonTablePrinter(TablePrinter):
+    def __init__(self, shell):
+        self._shell = shell
+        self._colnames = None
+        self._first_row = True
+
+    def print_header(self, formatted_names):
+        self._colnames = [n.strval for n in formatted_names]
+        self._shell.writeresult('[')
+
```
[the remark this answers is not in the document]
Comment (bschoening) [reply pylib/cqlshlib/displaying.py:210]: Duplicate column names shouldn't happen as they can lead to ambiguity and unexpected behavior.

**Comment by arvindKandpal-ksolves** [comment]
Hi @bschoening ,Please review the the latest changes , if anythink else I need to do , Please let me know.

**Comment by arvindKandpal-ksolves** [comment]
ping again @bschoening
